### PR TITLE
Setup database model oll 22

### PR DESCRIPTION
**Pull request by Toumi-Elyes**
## Summary

setup the database and the prisma orm and nest backend

## Checklist

**Bureaucracy**

- [x] I set a descriptive title
- [x] I set a proper description, providing enough information to the reviewer
- [x] I considered squashing this PR, if needed
- [x] I set the proper assignee
- [x] I set the proper reviewers

**Quality Assurance**

- [x] I tested the behavior of this PR locally and made sure it complies with
  the ticket description

### CODE REVIEW

File: backend/src/client.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { PrismaClient } from '@prisma/client';
+
+const prisma = new PrismaClient();
+
+async function main() {
+  console.log(`Database connected`);
```
Comment (ThisisYoYoDev) [inline backend/src/client.ts]: Why use injected string here ?

File: backend/src/user/user.service.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  validateEmail(email: string) {
+    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
+  }
+
+  hashPassword(password: string): string {
+    const hash = crypto.createHmac('sha512', Buffer.from(SECRET_KEY)).update(password).digest('hex');
+    return hash;
+  }
+
+  async postUser(userData: UserModel): Promise<string | null> {
```
Comment (ThisisYoYoDev) [inline backend/src/user/user.service.ts]: When you return null ?

File: docker-compose.yml [hunk 3]
```diff
@@ -1,20 +1,43 @@
...
-      POSTGRES_USER: ${POSTGRES_USER}
       POSTGRES_DB: ${POSTGRES_DB}
-    ports:
-      - ${POSTGRES_PORT}:${POSTGRES_PORT}
+      POSTGRES_USER: ${POSTGRES_USER}
+      POSTGRES_HOST: ${POSTGRES_HOST}
+    # networks:
+    #   - db
+
+  # backend:
```
Comment (ThisisYoYoDev) [inline docker-compose.yml]: Remove comment

**Review by ThisisYoYoDev — CHANGES_REQUESTED** [review]

File: backend/prisma/schema.prisma [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  admins    String[]
+  owner     String
+  Community Community[]
+  Course    Course[]
+}
+
+model Community {
+  id      String   @id @unique @default(uuid())
+  crew    Crew     @relation(references: [id], fields: [crew_id])
+  crew_id String
```
Comment (jeanbaptistedebize) [inline backend/prisma/schema.prisma]: inverse relation

File: backend/prisma/migrations/20221206172321_init/migration.sql [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+ALTER TABLE "Course" ADD CONSTRAINT "Course_crew_id_fkey" FOREIGN KEY ("crew_id") REFERENCES "Crew"("id") ON DELETE RESTRICT ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "Section" ADD CONSTRAINT "Section_course_id_fkey" FOREIGN KEY ("course_id") REFERENCES "Course"("id") ON DELETE RESTRICT ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "Chapter" ADD CONSTRAINT "Chapter_section_id_fkey" FOREIGN KEY ("section_id") REFERENCES "Section"("id") ON DELETE RESTRICT ON UPDATE CASCADE;
+
+-- AddForeignKey
+ALTER TABLE "Lessons" ADD CONSTRAINT "Lessons_chapter_id_fkey" FOREIGN KEY ("chapter_id") REFERENCES "Chapter"("id") ON DELETE RESTRICT ON UPDATE CASCADE;
```
Comment (jeanbaptistedebize) [inline backend/prisma/migrations/20221206172321_init/migration.sql]: Fix `on delete restrict` to `on delete cascade`.

File: backend/prisma/schema.prisma [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  provider = "prisma-client-js"
+}
+
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+
+model User {
+  id             String   @id @unique @default(uuid())
```
Comment (jeanbaptistedebize) [inline backend/prisma/schema.prisma]: remove unique. Id is already unique

File: backend/src/main.ts [hunk 7]
```diff
@@ -1,7 +1,11 @@
 import { NestFactory } from '@nestjs/core';
 import { SwaggerModule, DocumentBuilder } from '@nestjs/swagger';
-import { AppModule } from './app.module';
+import { AppModule } from 'app.module';
+import { BACKEND_PORT } from 'setup';
 import * as fs from 'fs';
+import * as cookieParser from 'cookie-parser';
```
Comment (jeanbaptistedebize) [inline backend/src/main.ts]: only import require module plz

File: backend/src/setup.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import * as env from 'env-var';
+import * as dotenv from 'dotenv';
```
Comment (jeanbaptistedebize) [inline backend/src/setup.ts]: same

File: backend/src/user/user.service.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    if (!token) {
+      Logger.error('Token not created !');
+      throw new ConflictException('Token not created !');
+    }
+
+    return token;
+  }
+
+  validateEmail(email: string) {
```
Comment (jeanbaptistedebize) [inline backend/src/user/user.service.ts]: should be private ?

File: backend/src/user/user.service.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      where: {
+        email: userData.Email,
+      },
+    });
+
+    if (user) {
+      Logger.error('Email already exists !');
+      throw new ConflictException('Email already exists !');
+    }
+
```
Comment (jeanbaptistedebize) [inline backend/src/user/user.service.ts]: move it in try catch here

**Review by jeanbaptistedebize — APPROVED** [review]

**Pull request merged**